### PR TITLE
Improve performance of wildcard filters

### PR DESCRIPTION
**Pull request by warwickmm**
This improves the performance of wildcard filters by avoiding unnecessary array allocations.  The mono profiler indicated that a large percentage of time was spent allocating arrays that were used as stacks.  By replacing these with `Stack` objects, we avoid this unnecessary overhead.

This also adds some tests to verify the behavior of the wildcard filter.

Note that this does not improve the performance of the default excludes filters.  To address that issue, we need to avoid converting them to a regex filter and preserve their wildcard nature.

### CODE REVIEW

File: Duplicati/Library/Utility/FilterExpression.cs [hunk 1]
```diff
@@ -186,9 +186,9 @@ private static Regex GetFilterGroupRegex(string filterGroupName)
...
             private static bool IsWildcardMatch(string input, string pattern)
             {
-                int[] inputPosStack = new int[(input.Length + 1) * (pattern.Length + 1)];   // Stack containing input positions that should be tested for further matching
-                int[] patternPosStack = new int[inputPosStack.Length];                      // Stack containing pattern positions that should be tested for further matching
-                int stackPos = -1;                                                          // Points to last occupied entry in stack; -1 indicates that stack is empty
+                Stack<int> inputPosStack = new Stack<int>(); // Stack containing input positions that should be tested for further matching
+                Stack<int> patternPosStack = new Stack<int>(); // Stack containing pattern positions that should be tested for further matching
+
                 bool[,] pointTested = new bool[input.Length + 1, pattern.Length + 1];       // Each true value indicates that input position vs. pattern position has been tested
```
Comment (warwickmm) [inline Duplicati/Library/Utility/FilterExpression.cs]: I made several attempts to replace this multidimensional array with various data structures (`HashSet<Tuple<int, int>>`, `bool[][]`), but all my attempts proved unsuccessful in improving the performance.  For some reason, this allocation does not seem to have as large of an effect on performance as the removed array allocations above.

**Comment by warwickmm** [comment]
Prior to version v2.0.4.8-2.0.4.8_canary_2018-12-13, enumerating files with the default filters would take 18-20x the time it took with no filters.  These changes, together with those from pull request #3545, should hopefully reduce the impact to around 1.5-2x.

**Comment by mikaelmello** [comment]
Wonderful work!

**Comment by warwickmm** [comment]
Unfortunately, I'm having difficulty realizing the performance gains in an actual backup configuration.  My targeted benchmarks definitely show some improvement (as described above), but attempts to observe this in an actual backup has proved more difficult.

I have some more testing I need to do before we consider merging this and will report back with what I find.

**Comment by mikaelmello** [comment]
Which tools do you use for benchmarking and how do you analyze performance on actual backups?

**Comment by warwickmm** [comment]
For benchmarking `Utility.EnumerateFileSystemEntries` with various filters, I was using [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet).  For the actual backup, I was just clicking Run in the UI and looking at the reported timings.

**Comment by warwickmm** [comment]
I think these changes are ok.  However, I would like to reword the commit message before we merge, as this does not address issue #3395.  Once I fix the message, I'll edit the pull request title accordingly.

It looks like the reason why this doesn't improve the performance of the default excludes is because the code converts all of those wildcard filters into one giant regex (possibly because the default excludes used to contain a regex filter).  I think the code already has support for treating these as a collection of wildcard filters.  I will try and look into that next.

**Pull request merged**